File: 27week/LeetCode-179-Largest-Number.py
import functools

class Solution:
    def largestNumber(self, nums: list[int]) -> str:
        str_nums = [str(num) for num in nums]

        def compare(x, y):
            if len(x) == len(y):
                return int(x) - int(y)
            elif len(x) > len(y):
                if x[:len(y)] == y:
                    return compare(x + y, y + x)
                else:
                    return int(x[:len(y)]) - int(y)
            else: 
                return -1 * compare(y, x)

        str_nums.sort(key=functools.cmp_to_key(compare), reverse=True)

        return ''.join(str_nums).lstrip('0') or '0'

File: 2024상반기/5월 4주차/leetcode-179-largest number.cpp
class Solution {
public:
    static bool comparator(int an, int bn) {
        string a = to_string(an);
        string b = to_string(bn);
        string ab = a + b;
        string ba = b + a;
        return stod(ab) > stod(ba);
    }
    string largestNumber(vector<int>& nums) {
        if (count(nums.begin(), nums.end(), 0) == nums.size()) return "0";
        sort(nums.begin(), nums.end(), comparator);
        string answer = "";
        for (int n : nums) answer += to_string(n);
        return answer;
    }
};